cpp/src/allgather.cpp

```diff
@@ -0,0 +1,149 @@
+ *
+ * The implementation creates a copy of the data for each peer during insertion and assign
+ * the peer's rank as the partition ID, i.e., there will only be one partition per rank.
+ * `partition_owner` lambda function is an identity function. These enable the data
+ * copies to be sent to the corresponding rank using the Shuffler.
+ *
+ * @note This implementation might not be scalable, as it makes a copy of the data for
+ * each peer during insertion.
+ */
+
+AllGather::AllGather(
+    std::shared_ptr<Communicator> comm,
+    std::shared_ptr<ProgressThread> progress_thread,
+    OpID op_id,
+    rmm::cuda_stream_view stream,
+    BufferResource* br,
+    std::shared_ptr<Statistics> statistics
+)
+    : comm_(comm.get()),
+      shuffler_(
+          std::make_unique<shuffler::Shuffler>(
+              std::move(comm),
+              std::move(progress_thread),
+              op_id,
+              comm_->nranks(),
+              stream,
+              br,
+              std::move(statistics),
+              [](std::shared_ptr<Communicator> const&, shuffler::PartID pid) -> Rank {
+                  // identity-like mapping, as there are only n_ranks partitions
+                  return static_cast<Rank>(pid);
+              }
+          )
+      ),
+      stream_(stream),
+      br_(br) {}
+
+AllGather::~AllGather() {
+    shutdown();
+}
+
+void AllGather::shutdown() {
+    if (shuffler_) {
+        shuffler_->shutdown();
+        shuffler_.reset();
+    }
+}
+
+void AllGather::insert(PackedData&& data) {
+    std::unordered_map<shuffler::PartID, PackedData> chunks;
+    chunks.reserve(static_cast<size_t>(comm_->nranks()));
+
+    for (Rank r = 0; r < comm_->nranks(); ++r) {
+        if (r == comm_->rank()) {
+            continue;
+        }
+        // copy data in the chunk
+        auto metadata_buf = std::make_unique<std::vector<uint8_t>>(*data.metadata);
+        auto res = reserve_or_fail(br_, data.data->size);
+        auto data_buf = br_->copy(data.data, stream_, res);
+        chunks.emplace(
+            static_cast<shuffler::PartID>(r),
+            PackedData{std::move(metadata_buf), std::move(data_buf)}
+        );
+    }
+    chunks.emplace(static_cast<shuffler::PartID>(comm_->rank()), std::move(data));
+
+    shuffler_->insert(std::move(chunks));
+}
+
+void AllGather::insert_finished() {
+    bool expected = false;
+    if (insert_finished_.compare_exchange_strong(expected, true)) {
+        // there will be n_ranks partitions
+        std::vector<shuffler::PartID> pids(static_cast<size_t>(comm_->nranks()));
+        std::iota(pids.begin(), pids.end(), 0);
+        shuffler_->insert_finished(std::move(pids));
+    }
+}
+
+bool AllGather::finished() const {
+    return shuffler_->finished();
+}
+
+std::vector<PackedData> AllGather::wait_and_extract(
+    std::optional<std::chrono::milliseconds> timeout
+) {
+    RAPIDSMPF_EXPECTS(
+        insert_finished_.load(), "insertion has not finished yet.", std::runtime_error
+    );
```

Comment:
This tracking locally whether I have finished inserting. But that is not equivalent to saying that everyone has finished inserting. So I think this flag should be removed because the error checking it provides is not correct.

Reply:
IOW, marking locally that I have finished insertions is a necessary *but not sufficient* condition for knowing that it is safe to extract data.